File: tests/wasm/test_dx.js
/**
 * WASM binding tests for DX indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('DX accuracy', () => {
    
    const high = testData.high;
    const low = testData.low;
    const close = testData.close;
    const expected = EXPECTED_OUTPUTS.dx;
    
    const result = wasm.dx_js(
        high, low, close,
        expected.defaultParams.period
    );
    
    assert.strictEqual(result.length, high.length, 'Output length should match input length');
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-4,  
        "DX last 5 values mismatch"
    );
    
    
    for (let i = 0; i < result.length; i++) {
        if (!isNaN(result[i])) {
            assert(result[i] >= 0 && result[i] <= 100, 
                `DX value at index ${i} should be between 0 and 100, got ${result[i]}`);
        }
    }
});

test('DX basic functionality', () => {
    const high = testData.high;
    const low = testData.low;
    const close = testData.close;
    
    
    const result = wasm.dx_js(high, low, close, 14);
    assert.strictEqual(result.length, high.length, 'Output length should match input length');
    
    
    const warmup = 14 - 1;
    for (let i = 0; i < warmup; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} during warmup period`);
    }
    
    
    let hasValidValues = false;
    for (let i = warmup + 10; i < result.length; i++) {
        if (!isNaN(result[i])) {
            hasValidValues = true;
            break;
        }
    }
    assert(hasValidValues, 'Should have valid values after warmup period');
});

test('DX warmup period validation', () => {
    const high = testData.high.slice(0, 100);
    const low = testData.low.slice(0, 100);
    const close = testData.close.slice(0, 100);
    
    const period = 14;
    const result = wasm.dx_js(high, low, close, period);
    
    
    
    
    const expectedWarmup = period - 1;
    
    
    let lastNaNIndex = -1;
    for (let i = 0; i < result.length; i++) {
        if (isNaN(result[i])) {
            lastNaNIndex = i;
        } else {
            break;
        }
    }
    
    assert(lastNaNIndex >= expectedWarmup - 1, 
        `Expected at least ${expectedWarmup} NaN values, but last NaN was at index ${lastNaNIndex}`);
    
    
    if (result.length > expectedWarmup + 10) {
        for (let i = expectedWarmup + 10; i < Math.min(expectedWarmup + 20, result.length); i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} after warmup period`);
        }
    }
});

test('DX value range validation', () => {
    const high = testData.high;
    const low = testData.low;
    const close = testData.close;
    
    
    const periods = [5, 10, 14, 20, 50];
    
    for (const period of periods) {
        if (period >= high.length) continue;
        
        const result = wasm.dx_js(high, low, close, period);
        
        
        let minVal = Infinity;
        let maxVal = -Infinity;
        let validCount = 0;
        
        for (let i = 0; i < result.length; i++) {
            if (!isNaN(result[i])) {
                validCount++;
                minVal = Math.min(minVal, result[i]);
                maxVal = Math.max(maxVal, result[i]);
                
                assert(result[i] >= -1e-9, 
                    `DX value at index ${i} should be >= 0, got ${result[i]} for period ${period}`);
                assert(result[i] <= 100.0 + 1e-9, 
                    `DX value at index ${i} should be <= 100, got ${result[i]} for period ${period}`);
            }
        }
        
        
        assert(validCount > 0, `No valid values found for period ${period}`);
        
        console.log(`Period ${period}: min=${minVal.toFixed(2)}, max=${maxVal.toFixed(2)}, valid=${validCount}`);
    }
});

test('DX fast API', () => {
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low); 
    const close = new Float64Array(testData.close);
    const len = high.length;
    
    
    const highPtr = wasm.dx_alloc(len);
    const lowPtr = wasm.dx_alloc(len);
    const closePtr = wasm.dx_alloc(len);
    const outPtr = wasm.dx_alloc(len);
    
    try {
        
        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);
        highMem.set(high);
        lowMem.set(low);
        closeMem.set(close);
        
        
        wasm.dx_into(
            highPtr,
            lowPtr,
            closePtr,
            outPtr,
            len,
            14
        );
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        const resultCopy = Float64Array.from(result);
        
        
        const safeResult = wasm.dx_js(high, low, close, 14);
        assertArrayClose(resultCopy, safeResult, 1e-10, 'Fast API should match safe API');
        
    } finally {
        wasm.dx_free(highPtr, len);
        wasm.dx_free(lowPtr, len);
        wasm.dx_free(closePtr, len);
        wasm.dx_free(outPtr, len);
    }
});

test('DX fast API with aliasing', () => {
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const len = high.length;
    
    
    const highPtr = wasm.dx_alloc(len);
    const lowPtr = wasm.dx_alloc(len);
    const closePtr = wasm.dx_alloc(len);
    
    try {
        
        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);
        highMem.set(high);
        lowMem.set(low);
        closeMem.set(close);
        
        
        wasm.dx_into(
            highPtr,
            lowPtr,
            closePtr,
            highPtr,  
            len,
            14
        );
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const resultCopy = Float64Array.from(result);
        
        
        const expected = wasm.dx_js(high, low, close, 14);
        assertArrayClose(resultCopy, expected, 1e-10, 'Should handle aliasing correctly');
        
    } finally {
        wasm.dx_free(highPtr, len);
        wasm.dx_free(lowPtr, len);
        wasm.dx_free(closePtr, len);
    }
});

test('DX batch API with unified interface', () => {
    const high = testData.high.slice(0, 100);  
    const low = testData.low.slice(0, 100);
    const close = testData.close.slice(0, 100);
    
    const config = {
        period_range: [10, 20, 5]  
    };
    
    const result = wasm.dx_batch(high, low, close, config);
    
    
    assert(result.values, 'Result should have values array');
    assert(result.combos, 'Result should have combos array');
    assert.strictEqual(result.rows, 3, 'Should have 3 parameter combinations');
    assert.strictEqual(result.cols, high.length, 'Cols should match input length');
    assert.strictEqual(result.values.length, 3 * high.length, 'Values should be flattened matrix');
    
    
    const periods = [10, 15, 20];
    for (let i = 0; i < periods.length; i++) {
        const batchRow = result.values.slice(i * high.length, (i + 1) * high.length);
        const singleResult = wasm.dx_js(high, low, close, periods[i]);
        assertArrayClose(batchRow, singleResult, 1e-10, 
            `Batch result for period ${periods[i]} should match single computation`);
    }
});

test('DX batch with single parameter', () => {
    const high = testData.high;
    const low = testData.low;
    const close = testData.close;
    
    const config = {
        period_range: [14, 14, 0]  
    };
    
    const result = wasm.dx_batch(high, low, close, config);
    
    
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, high.length);
    
    
    const batchRow = result.values;
    const expected = EXPECTED_OUTPUTS.dx.last5Values;
    
    
    const last5 = batchRow.slice(-5);
    assertArrayClose(
        last5,
        expected,
        1e-4,
        "DX batch default row mismatch"
    );
});

test('DX batch metadata', () => {
    const high = new Float64Array(50);
    const low = new Float64Array(50);
    const close = new Float64Array(50);
    
    
    for (let i = 0; i < 50; i++) {
        high[i] = 100 + Math.sin(i * 0.1) * 10;
        low[i] = 90 + Math.sin(i * 0.1) * 10;
        close[i] = 95 + Math.sin(i * 0.1) * 10;
    }
    
    const config = {
        period_range: [10, 20, 10]  
    };
    
    const result = wasm.dx_batch(high, low, close, config);
    
    
    assert.strictEqual(result.combos.length, 2);
    
    
    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[1].period, 20);
    
    
    const row1 = result.values.slice(0, 50);
    const row2 = result.values.slice(50, 100);
    
    
    let nanCount1 = 0;
    for (let i = 0; i < 15; i++) {
        if (isNaN(row1[i])) nanCount1++;
    }
    
    
    let nanCount2 = 0;
    for (let i = 0; i < 25; i++) {
        if (isNaN(row2[i])) nanCount2++;
    }
    
    assert(nanCount1 >= 9, `Expected at least 9 NaN values for period 10, got ${nanCount1}`);
    assert(nanCount2 >= 19, `Expected at least 19 NaN values for period 20, got ${nanCount2}`);
});

test('DX error handling', () => {
    const high = testData.high;
    const low = testData.low;
    const close = testData.close;
    
    
    assert.throws(() => {
        wasm.dx_js([], [], [], 14);
    }, 'Should throw on empty input');
    
    
    const result = wasm.dx_js(high.slice(0, 10), low.slice(0, 5), close.slice(0, 10), 2);
    assert.strictEqual(result.length, 5, 'Should use minimum input length');
    
    
    assert.throws(() => {
        wasm.dx_js(high.slice(0, 10), low.slice(0, 10), close.slice(0, 10), 20);
    }, 'Should throw when period exceeds data length');
    
    
    assert.throws(() => {
        wasm.dx_js(high, low, close, 0);
    }, 'Should throw on zero period');
});

test('DX all NaN input', () => {
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.dx_js(allNaN, allNaN, allNaN, 14);
    }, /All high, low, and close values are NaN/, 'Should throw when all values are NaN');
});

test('DX batch edge cases', () => {
    const high = new Float64Array([100, 102, 101, 103, 104, 105, 103, 102, 104, 106]);
    const low = new Float64Array([98, 99, 98, 100, 101, 102, 100, 99, 101, 103]);
    const close = new Float64Array([99, 101, 100, 102, 103, 104, 102, 101, 103, 105]);
    
    
    const singleBatch = wasm.dx_batch(high, low, close, {
        period_range: [5, 5, 1]
    });
    
    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.combos.length, 1);
    
    
    const largeBatch = wasm.dx_batch(high, low, close, {
        period_range: [5, 7, 10]  
    });
    
    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.combos.length, 1);
    assert.strictEqual(largeBatch.combos[0].period, 5);
    
    
    assert.throws(() => {
        wasm.dx_batch([], [], [], {
            period_range: [14, 14, 0]
        });
    }, 'Should throw on empty data');
});

test('DX memory management', () => {
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.dx_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 2.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 2.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.dx_free(ptr, size);
    }
});

test.after(() => {
    console.log('DX WASM tests completed');
});

File: tests/python/test_dx.py
"""
Python binding tests for DX indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestDx:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_dx_accuracy(self, test_data):
        """Test DX matches expected values from Rust tests - mirrors check_dx_accuracy"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['dx']
        
        result = ta_indicators.dx(
            high, low, close,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-4,  # DX uses less precise tolerance
            msg="DX last 5 values mismatch"
        )
        
        # Verify DX values are in valid range [0, 100]
        valid_values = result[~np.isnan(result)]
        assert np.all((valid_values >= 0) & (valid_values <= 100)), \
            "DX values should be between 0 and 100"
    
    def test_dx_default_params(self, test_data):
        """Test DX with default parameters - mirrors check_dx_default_candles"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        # Default period is 14
        result = ta_indicators.dx(high, low, close, period=14)
        assert len(result) == len(close)
        
        # Verify warmup period (should have NaN values)
        # Warmup = first_valid_idx + period - 1 (minimum 13)
        assert np.any(np.isnan(result[:13])), "Expected NaN in warmup period"
        
        # Verify we have valid values after warmup
        if len(result) > 50:
            assert not np.all(np.isnan(result[50:])), "Expected valid values after warmup"
    
    def test_dx_zero_period(self, test_data):
        """Test DX fails with zero period - mirrors check_dx_zero_period"""
        high = np.array([2.0, 2.5, 3.0])
        low = np.array([1.0, 1.2, 2.1])
        close = np.array([1.5, 2.3, 2.2])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.dx(high, low, close, period=0)
    
    def test_dx_period_exceeds_length(self):
        """Test DX fails when period exceeds data length - mirrors check_dx_period_exceeds_length"""
        high = np.array([3.0, 4.0])
        low = np.array([2.0, 3.0])
        close = np.array([2.5, 3.5])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.dx(high, low, close, period=14)
    
    def test_dx_very_small_dataset(self):
        """Test DX fails with insufficient data - mirrors check_dx_very_small_dataset"""
        high = np.array([3.0])
        low = np.array([2.0])
        close = np.array([2.5])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.dx(high, low, close, period=14)
    
    def test_dx_empty_input(self):
        """Test DX fails with empty input - mirrors check_dx_empty_input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.dx(empty, empty, empty, period=14)
    
    def test_dx_all_nan_input(self):
        """Test DX with all NaN values - mirrors check_dx_all_nan"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All high, low, and close values are NaN"):
            ta_indicators.dx(all_nan, all_nan, all_nan, period=14)
    
    def test_dx_mismatched_lengths(self):
        """Test DX fails with mismatched input lengths"""
        high = np.array([1.0, 2.0, 3.0])
        low = np.array([1.0, 2.0])  # Different length
        close = np.array([1.0, 2.0, 3.0])
        
        # The function should handle this gracefully by using min length
        result = ta_indicators.dx(high, low, close, period=1)
        assert len(result) == 2  # min(3, 2, 3) = 2
    
    def test_dx_nan_handling(self, test_data):
        """Test DX handles NaN values correctly - mirrors check_dx_nan_handling"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        result = ta_indicators.dx(high, low, close, period=14)
        assert len(result) == len(close)
        
        # After sufficient warmup period, no NaN values should exist
        if len(result) > 50:
            assert not np.any(np.isnan(result[50:])), \
                "Found unexpected NaN after warmup period"
        
        # First period-1 values should be NaN (minimum 13 for period=14)
        assert np.any(np.isnan(result[:13])), "Expected NaN in warmup period"
    
    def test_dx_streaming(self, test_data):
        """Test DX streaming matches batch calculation - mirrors check_dx_streaming"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        period = 14
        
        # Batch calculation
        batch_result = ta_indicators.dx(high, low, close, period=period)
        
        # Streaming calculation
        stream = ta_indicators.DxStream(period=period)
        stream_values = []
        
        for h, l, c in zip(high, low, close):
            result = stream.update(h, l, c)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        # Compare batch vs streaming
        assert len(batch_result) == len(stream_values)
        
        # Compare values where both are not NaN
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"DX streaming mismatch at index {i}")
    
    def test_dx_batch_single_params(self, test_data):
        """Test DX batch processing with single parameter - mirrors check_batch_default_row"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        result = ta_indicators.dx_batch(
            high, low, close,
            period_range=(14, 14, 0)  # Default period only
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 1 combination
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        # Extract the single row
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['dx']['last_5_values']
        
        # Check last 5 values match
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-4,
            msg="DX batch default row mismatch"
        )
    
    def test_dx_batch_multiple_periods(self, test_data):
        """Test DX batch with multiple periods - mirrors check_batch_sweep"""
        high = test_data['high'][:100]  # Use smaller dataset for speed
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        
        result = ta_indicators.dx_batch(
            high, low, close,
            period_range=(10, 30, 5)  # periods: 10, 15, 20, 25, 30
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 5 combinations
        expected_combos = 5
        assert result['values'].shape[0] == expected_combos
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == expected_combos
        
        # Verify periods array
        expected_periods = [10, 15, 20, 25, 30]
        np.testing.assert_array_equal(result['periods'], expected_periods)
        
        # Verify each batch result matches individual computation
        for i, period in enumerate(expected_periods):
            batch_row = result['values'][i]
            single_result = ta_indicators.dx(high, low, close, period=period)
            assert_close(batch_row, single_result, rtol=1e-10,
                        msg=f"Batch result for period {period} should match single computation")
    
    def test_dx_batch_with_kernel(self, test_data):
        """Test DX batch with kernel parameter"""
        high = test_data['high'][:50]
        low = test_data['low'][:50]
        close = test_data['close'][:50]
        
        # Test with different kernels
        result_auto = ta_indicators.dx_batch(
            high, low, close,
            period_range=(14, 14, 0),
            kernel=None  # Auto kernel
        )
        
        result_scalar = ta_indicators.dx_batch(
            high, low, close,
            period_range=(14, 14, 0),
            kernel='scalar'
        )
        
        # Both should produce valid results
        assert result_auto['values'].shape == result_scalar['values'].shape
        
        # Results should be very close (might differ slightly due to kernel differences)
        assert_close(
            result_auto['values'][0],
            result_scalar['values'][0],
            rtol=1e-8,
            msg="Different kernels should produce similar results"
        )
    
    def test_dx_value_bounds(self, test_data):
        """Test that DX values are always between 0 and 100"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        # Test with various periods
        for period in [5, 10, 14, 20, 50]:
            if period >= len(high):
                continue
                
            result = ta_indicators.dx(high, low, close, period=period)
            
            # Check all non-NaN values are in valid range
            valid_values = result[~np.isnan(result)]
            if len(valid_values) > 0:
                assert np.all(valid_values >= -1e-9), \
                    f"DX values should be >= 0 for period {period}"
                assert np.all(valid_values <= 100.0 + 1e-9), \
                    f"DX values should be <= 100 for period {period}"
    
    def test_dx_with_kernel_parameter(self, test_data):
        """Test DX with explicit kernel parameter"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        
        # Test with scalar kernel
        result_scalar = ta_indicators.dx(
            high, low, close,
            period=14,
            kernel='scalar'
        )
        
        # Test with auto kernel
        result_auto = ta_indicators.dx(
            high, low, close,
            period=14,
            kernel=None
        )
        
        # Both should have same length
        assert len(result_scalar) == len(result_auto) == len(close)
        
        # Results should be very close
        assert_close(
            result_scalar,
            result_auto,
            rtol=1e-8,
            msg="Different kernels should produce similar results"
        )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])